File: FLYR/Features/Stats/Models/UserStats.swift
import Foundation

struct UserStats: Codable, Identifiable {
    let id: UUID
    let user_id: UUID
    let day_streak: Int
    let best_streak: Int
    let doors_knocked: Int
    let flyers: Int
    let conversations: Int
    let leads_created: Int
    let appointments: Int
    let qr_codes_scanned: Int
    let distance_walked: Double
    let time_tracked: Int
    let conversation_per_door: Double
    let conversation_lead_rate: Double
    let qr_code_scan_rate: Double
    let qr_code_lead_rate: Double
    let streak_days: [String]?
    let xp: Int
    let updated_at: String
    let created_at: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case user_id
        case day_streak
        case best_streak
        case doors_knocked
        case flyers
        case conversations
        case leads_created
        case appointments
        case qr_codes_scanned
        case distance_walked
        case time_tracked
        case conversation_per_door
        case conversation_lead_rate
        case qr_code_scan_rate
        case qr_code_lead_rate
        case streak_days
        case xp
        case updated_at
        case created_at
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        user_id = try container.decode(UUID.self, forKey: .user_id)
        day_streak = try container.decode(Int.self, forKey: .day_streak)
        best_streak = try container.decode(Int.self, forKey: .best_streak)
        doors_knocked = try container.decode(Int.self, forKey: .doors_knocked)
        flyers = try container.decode(Int.self, forKey: .flyers)
        conversations = try container.decode(Int.self, forKey: .conversations)
        leads_created = try container.decode(Int.self, forKey: .leads_created)
        appointments = (try? container.decodeIfPresent(Int.self, forKey: .appointments)) ?? 0
        qr_codes_scanned = try container.decode(Int.self, forKey: .qr_codes_scanned)
        distance_walked = try container.decode(Double.self, forKey: .distance_walked)
        time_tracked = try container.decode(Int.self, forKey: .time_tracked)
        conversation_per_door = try container.decode(Double.self, forKey: .conversation_per_door)
        conversation_lead_rate = try container.decode(Double.self, forKey: .conversation_lead_rate)
        qr_code_scan_rate = try container.decode(Double.self, forKey: .qr_code_scan_rate)
        qr_code_lead_rate = try container.decode(Double.self, forKey: .qr_code_lead_rate)
        streak_days = try container.decodeIfPresent([String].self, forKey: .streak_days)
        xp = try container.decode(Int.self, forKey: .xp)
        updated_at = try container.decode(String.self, forKey: .updated_at)
        created_at = try container.decodeIfPresent(String.self, forKey: .created_at)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(user_id, forKey: .user_id)
        try container.encode(day_streak, forKey: .day_streak)
        try container.encode(best_streak, forKey: .best_streak)
        try container.encode(doors_knocked, forKey: .doors_knocked)
        try container.encode(flyers, forKey: .flyers)
        try container.encode(conversations, forKey: .conversations)
        try container.encode(leads_created, forKey: .leads_created)
        try container.encode(appointments, forKey: .appointments)
        try container.encode(qr_codes_scanned, forKey: .qr_codes_scanned)
        try container.encode(distance_walked, forKey: .distance_walked)
        try container.encode(time_tracked, forKey: .time_tracked)
        try container.encode(conversation_per_door, forKey: .conversation_per_door)
        try container.encode(conversation_lead_rate, forKey: .conversation_lead_rate)
        try container.encode(qr_code_scan_rate, forKey: .qr_code_scan_rate)
        try container.encode(qr_code_lead_rate, forKey: .qr_code_lead_rate)
        try container.encodeIfPresent(streak_days, forKey: .streak_days)
        try container.encode(xp, forKey: .xp)
        try container.encode(updated_at, forKey: .updated_at)
        try container.encodeIfPresent(created_at, forKey: .created_at)
    }
    
    // Helper computed properties for display
    var formattedDistanceWalked: String {
        String(format: "%.1f", distance_walked)
    }
    
    var formattedConversationPerDoor: String {
        String(format: "%.1f", conversation_per_door)
    }
    
    var formattedTimeTracked: String {
        let hours = time_tracked / 60
        let minutes = time_tracked % 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        }
        return "\(minutes)m"
    }
}

File: web/src/lib/userStats.ts
import { supabase } from '../supabase'
import type { UserStats } from '../types/userStats'

type StatsRow = Record<string, unknown>

function numberValue(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0
}

function leadKey(row: StatsRow): string {
  const address = String(row.address ?? '').trim().toLowerCase()
  return address || String(row.id ?? '').toLowerCase()
}

function appointmentKey(row: StatsRow): string {
  const timestamp = Date.parse(String(row.timestamp ?? ''))
  const seconds = Number.isFinite(timestamp) ? Math.round(timestamp / 1000) : 0
  return [
    String(row.contact_id ?? '').toLowerCase(),
    String(seconds),
    String(row.note ?? '').trim().toLowerCase(),
  ].join('|')
}

function defaultStats(userId: string): UserStats {
  return {
    id: '',
    user_id: userId,
    day_streak: 0,
    best_streak: 0,
    doors_knocked: 0,
    flyers: 0,
    conversations: 0,
    leads_created: 0,
    appointments: 0,
    qr_codes_scanned: 0,
    distance_walked: 0,
    time_tracked: 0,
    conversation_per_door: 0,
    conversation_lead_rate: 0,
    qr_code_scan_rate: 0,
    qr_code_lead_rate: 0,
    streak_days: null,
    xp: 0,
    updated_at: new Date().toISOString(),
    created_at: null,
  }
}

function mapRow(row: Record<string, unknown>): UserStats {
  return {
    id: String(row.id ?? ''),
    user_id: String(row.user_id ?? ''),
    day_streak: numberValue(row.day_streak),
    best_streak: numberValue(row.best_streak),
    doors_knocked: numberValue(row.doors_knocked),
    flyers: numberValue(row.flyers),
    conversations: numberValue(row.conversations),
    leads_created: numberValue(row.leads_created),
    appointments: numberValue(row.appointments),
    qr_codes_scanned: numberValue(row.qr_codes_scanned),
    distance_walked: numberValue(row.distance_walked),
    time_tracked: numberValue(row.time_tracked),
    conversation_per_door: numberValue(row.conversation_per_door),
    conversation_lead_rate: numberValue(row.conversation_lead_rate),
    qr_code_scan_rate: numberValue(row.qr_code_scan_rate),
    qr_code_lead_rate: numberValue(row.qr_code_lead_rate),
    streak_days: Array.isArray(row.streak_days) ? (row.streak_days as string[]) : null,
    xp: numberValue(row.xp),
    updated_at: String(row.updated_at ?? ''),
    created_at: row.created_at != null ? String(row.created_at) : null,
  }
}

async function refreshUserStatsFromSessions(userId: string): Promise<void> {
  if (!supabase) return

  const { error } = await supabase.rpc('refresh_user_stats_from_sessions', {
    p_user_id: userId,
  })
  if (error) {
    // Keep the page usable if an older environment has not deployed the RPC yet.
    console.warn('Failed to refresh user stats from sessions', error)
  }
}

async function fetchLiveLeadCount(userId: string): Promise<number> {
  if (!supabase) return 0

  const leadKeys = new Set<string>()
  const { data: contacts, error: contactsError } = await supabase
    .from('contacts')
    .select('id,address')
    .eq('user_id', userId)
  if (contactsError) throw contactsError
  ;((contacts ?? []) as StatsRow[]).forEach((row) => leadKeys.add(leadKey(row)))

  const { data: legacyLeads, error: legacyLeadsError } = await supabase
    .from('field_leads')
    .select('id,address')
    .eq('user_id', userId)
  if (legacyLeadsError) throw legacyLeadsError
  ;((legacyLeads ?? []) as StatsRow[]).forEach((row) => leadKeys.add(leadKey(row)))

  return leadKeys.size
}

async function fetchLiveAppointmentCount(userId: string): Promise<number> {
  if (!supabase) return 0

  const appointmentKeys = new Set<string>()
  const { data, error } = await supabase
    .from('contact_activities')
    .select('id,contact_id,note,timestamp,contacts!inner(id,user_id)')
    .eq('type', 'meeting')
    .eq('contacts.user_id', userId)
  if (error) throw error
  ;((data ?? []) as StatsRow[]).forEach((row) => appointmentKeys.add(appointmentKey(row)))

  return appointmentKeys.size
}

async function mergedWithLiveLeadAndAppointmentCounts(stats: UserStats, userId: string): Promise<UserStats> {
  const [liveLeadCount, liveAppointmentCount] = await Promise.all([
    fetchLiveLeadCount(userId),
    fetchLiveAppointmentCount(userId),
  ])
  const leadsCreated = Math.max(stats.leads_created, liveLeadCount)
  const appointments = Math.max(stats.appointments, liveAppointmentCount)

  return {
    ...stats,
    leads_created: leadsCreated,
    appointments,
    conversation_lead_rate:
      stats.conversations > 0 ? leadsCreated / stats.conversations : stats.conversation_lead_rate,
  }
}

/**
 * Fetches the current user's stats from public.user_stats, refreshed from sessions
 * and merged with live lead/appointment counts to match the iOS stats surface.
 */
export async function fetchUserStats(userId: string): Promise<UserStats> {
  if (!supabase) {
    throw new Error('Supabase client not configured')
  }
  await refreshUserStatsFromSessions(userId)

  const { data, error } = await supabase
    .from('user_stats')
    .select()
    .eq('user_id', userId)
    .limit(1)
    .maybeSingle()
  if (error) throw error
  const base = data ? mapRow(data as Record<string, unknown>) : defaultStats(userId)
  return mergedWithLiveLeadAndAppointmentCounts(base, userId)
}
